{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Posting the Dockerfile which produced this jupyter image - modified off of https://github.com/DataBiosphere/terra-docker/blob/master/terra-jupyter-aou/Dockerfile\n",
    "# Primary change is adding a GATK bundle which has the tasks in the ah_var_store branch\n",
    "\n",
    "FROM us.gcr.io/broad-dsp-gcr-public/terra-jupyter-python:0.0.14 AS python\n",
    "\n",
    "FROM us.gcr.io/broad-dsp-gcr-public/terra-jupyter-r:1.0.4\n",
    "\n",
    "# copy everything pip installed from the python image\n",
    "COPY --from=python /usr/local/lib/python3.7/dist-packages /usr/local/lib/python3.7/dist-packages\n",
    "\n",
    "USER root\n",
    "\n",
    "# need to apt-get everything for python since we can only copy pip installed packages\n",
    "RUN apt-get update && apt-get install -yq --no-install-recommends \\\n",
    "  jq \\\n",
    "  python3.7-dev \\\n",
    "  python-tk \\\n",
    "  openjdk-8-jdk \\\n",
    "  tk-dev \\\n",
    "  libssl-dev \\\n",
    "  xz-utils \\\n",
    "  libhdf5-dev \\\n",
    "  openssl \\\n",
    "  make \\\n",
    "  liblzo2-dev \\\n",
    "  zlib1g-dev \\\n",
    "  libz-dev \\\n",
    "  libmagick++-dev \\\n",
    "  iproute2 \\\n",
    "  bcftools \\\n",
    "  # specify Java 8\n",
    "  && update-alternatives --set java /usr/lib/jvm/java-8-openjdk-amd64/jre/bin/java \\\n",
    "  && apt-get clean \\\n",
    "  && rm -rf /var/lib/apt/lists/*\n",
    "\n",
    "# Install Wondershaper from source, for client-side egress limiting.\n",
    "RUN cd /usr/local/share && \\\n",
    "  git clone https://github.com/magnific0/wondershaper.git --depth 1 && \\\n",
    "  cd wondershaper && \\\n",
    "  make install && \\\n",
    "  cd $HOME\n",
    "\n",
    "RUN echo \"Sys.setenv(RETICULATE_PYTHON = '$(which python3)')\" >> ~/.Rprofile\n",
    "\n",
    "# Install GATK\n",
    "ENV GATK_VERSION 4.1.7.0-31-g5ae5727-SNAPSHOT\n",
    "ENV GATK_ZIP_PATH /tmp/gatk.zip\n",
    "\n",
    "COPY gatk-ah_var_store-0.2.zip $GATK_ZIP_PATH\n",
    "RUN unzip -o $GATK_ZIP_PATH -d /etc/ \\\n",
    " && ln -s /etc/gatk-$GATK_VERSION/gatk /bin/gatk\n",
    "\n",
    "# Install Notebook libraries as the user.\n",
    "\n",
    "ENV USER jupyter-user\n",
    "USER $USER\n",
    "\n",
    "RUN pip3 install --upgrade \\\n",
    "  # Fix plotnine and its dependency mizani to a version that's compatible with 0.25.* of pandas.\n",
    "  # The next versions require pandas 1.0 which we're not ready to upgrade to just yet.\n",
    "  plotnine==0.6.0 \\\n",
    "  mizani==0.6.0 \\\n",
    "  statsmodels==0.10.0rc2 \\\n",
    "  \"git+git://github.com/all-of-us/workbench-snippets.git#egg=terra_widgets&subdirectory=py\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#The task will take ~6 hours so this is needed to prevent your cluster from pausing\n",
    "\n",
    "# This code snippet allows for modification of the autopause behavior for the current notebook server. This is\n",
    "# a stop-gap approach until we have integration in the Workbench/Notebooks UI to change the autopause settings\n",
    "# for notebook servers. *Warning*: an unpaused server incurs much higher cost and may exhaust free credits rapidly.\n",
    "# This setting is not entirely sticky. In production, it will be lost after 1-2w, depending on activity and when\n",
    "# the server was created.# To disable autopause, set it to false\n",
    "# To change the autopause threshold for how long to wait after last UI activity before\n",
    "# autopausing (unit: minutes), set autopauseThreshold (default is 30m)\n",
    "!curl -X PATCH \\\n",
    "  --header 'Content-Type: application/json' \\\n",
    "  --header \"Authorization: Bearer $(gcloud auth application-default print-access-token)\" \\\n",
    "  -d '{ \"autopause\": false, \"autopauseThreshold\": 60 }' \\\n",
    "  \"https://notebooks.firecloud.org/api/cluster/${GOOGLE_PROJECT}/${CLUSTER_NAME}\" \\\n",
    "  | jq '.'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Copying gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.dict...\n",
      "/ [0 files][    0.0 B/ 14.5 KiB]                                                \r",
      "/ [1 files][ 14.5 KiB/ 14.5 KiB]                                                \r",
      "Copying gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.fasta...\n",
      "/ [1 files][ 14.5 KiB/  2.9 GiB]                                                \r",
      "==> NOTE: You are downloading one or more large file(s), which would\n",
      "run significantly faster if you enabled sliced object downloads. This\n",
      "feature is enabled by default but requires that compiled crcmod be\n",
      "installed (see \"gsutil help crcmod\").\n",
      "\n",
      "-\r",
      "- [1 files][109.3 MiB/  2.9 GiB]                                                \r",
      "\\\r",
      "|\r",
      "| [1 files][245.7 MiB/  2.9 GiB]                                                \r",
      "/\r",
      "-\r",
      "- [1 files][376.2 MiB/  2.9 GiB]                                                \r",
      "\\\r",
      "\\ [1 files][492.4 MiB/  2.9 GiB]                                                \r",
      "|\r",
      "/\r",
      "/ [1 files][620.0 MiB/  2.9 GiB]                                                \r",
      "-\r",
      "\\\r",
      "\\ [1 files][751.3 MiB/  2.9 GiB]                                                \r",
      "|\r",
      "| [1 files][882.2 MiB/  2.9 GiB]                                                \r",
      "/\r",
      "-\r",
      "- [1 files][ 1014 MiB/  2.9 GiB]                                                \r",
      "\\\r",
      "|\r",
      "| [1 files][  1.1 GiB/  2.9 GiB]                                                \r",
      "/\r",
      "/ [1 files][  1.2 GiB/  2.9 GiB]                                                \r",
      "-\r",
      "\\\r",
      "\\ [1 files][  1.4 GiB/  2.9 GiB]  131.4 MiB/s                                   \r",
      "|\r",
      "/\r",
      "/ [1 files][  1.5 GiB/  2.9 GiB]  130.7 MiB/s                                   \r",
      "-\r",
      "- [1 files][  1.6 GiB/  2.9 GiB]  127.3 MiB/s                                   \r",
      "\\\r",
      "|\r",
      "| [1 files][  1.7 GiB/  2.9 GiB]  125.4 MiB/s                                   \r",
      "/\r",
      "-\r",
      "- [1 files][  1.9 GiB/  2.9 GiB]  125.2 MiB/s                                   \r",
      "\\\r",
      "\\ [1 files][  2.0 GiB/  2.9 GiB]  120.2 MiB/s                                   \r",
      "|\r",
      "/\r",
      "/ [1 files][  2.1 GiB/  2.9 GiB]  120.9 MiB/s                                   \r",
      "-\r",
      "\\\r",
      "\\ [1 files][  2.2 GiB/  2.9 GiB]  124.7 MiB/s                                   \r",
      "|\r",
      "| [1 files][  2.4 GiB/  2.9 GiB]  126.4 MiB/s                                   \r",
      "/\r",
      "-\r",
      "- [1 files][  2.5 GiB/  2.9 GiB]  126.6 MiB/s                                   \r",
      "\\\r",
      "|\r",
      "| [1 files][  2.6 GiB/  2.9 GiB]  131.0 MiB/s                                   \r",
      "/\r",
      "/ [1 files][  2.7 GiB/  2.9 GiB]  130.2 MiB/s                                   \r",
      "-\r",
      "\\\r",
      "\\ [1 files][  2.8 GiB/  2.9 GiB]  114.6 MiB/s                                   \r",
      "|\r",
      "/\r",
      "/ [1 files][  2.9 GiB/  2.9 GiB]  114.1 MiB/s                                   \r",
      "/ [2 files][  2.9 GiB/  2.9 GiB]      0.0 B/s                                   \r",
      "-\r",
      "Copying gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.fasta.fai...\n",
      "- [2 files][  2.9 GiB/  2.9 GiB]      0.0 B/s                                   \r",
      "- [3 files][  2.9 GiB/  2.9 GiB]   25.2 KiB/s                                   \r",
      "Copying gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/probe_info.csv...\n",
      "- [3 files][  2.9 GiB/  3.0 GiB]   25.2 KiB/s                                   \r",
      "\\\r",
      "\\ [4 files][  3.0 GiB/  3.0 GiB]   83.1 MiB/s                                   \r\n",
      "Operation completed over 4 objects/3.0 GiB.                                      \n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "gsutil cp 'gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.dict' .\n",
    "gsutil cp 'gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.fasta' .\n",
    "gsutil cp 'gs://fc-secure-36ff2d2c-861e-4a9f-b04a-6f08bb178679/data/extract/Homo_sapiens_assembly19.fasta.fai' ."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "PROJECT_ID=fc-aou-cdr-synth-test\n",
    "DATASET_NAME=microarray_data\n",
    "TABLE_NAME=probe_info\n",
    "\n",
    "OUTPUT_VCF=\"aou.vcf\"\n",
    "REF=\"Homo_sapiens_assembly19.fasta\"\n",
    "\n",
    "EXTRACT_PROBE_CLAUSE=\"--probe-info-csv probe_info.csv\"\n",
    "\n",
    "gatk ArrayExtractCohort -R \"${REF}\" \\\n",
    "   -O $OUTPUT_VCF \\\n",
    "   --use-compressed-data \"true\" \\\n",
    "   --cohort-extract-table $PROJECT_ID.$DATASET_NAME.arrays_003 \\\n",
    "   --local-sort-max-records-in-ram 10000000 \\\n",
    "   --sample-info-table $PROJECT_ID.$DATASET_NAME.sample_list \\\n",
    "   $EXTRACT_PROBE_CLAUSE \\\n",
    "   --project-id $PROJECT_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "bgzip -c aou.vcf > aou.vcf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Discrepancy in summary stat 'number of records:': Expected 1909964, Actual number of records:\n"
     ]
    },
    {
     "ename": "SystemExit",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "An exception has occurred, use %tb to see the full traceback.\n",
      "\u001b[0;31mSystemExit\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "import argparse\n",
    "import subprocess\n",
    "import sys\n",
    "import time\n",
    "\n",
    "# Given the 'expected' cohort vcf and the actual cohort VCF, we want to check four things:\n",
    "# 1) The VCFs have the same number of variants\n",
    "# 2) The VCFs have the same number of samples\n",
    "# 3) The VCFs have the same allele frequency across sites (with fuzziness for float variance)\n",
    "# 4) All genotypes match for, say, the first sample in the cohort\n",
    "start = time.time()\n",
    "#parser = argparse.ArgumentParser(description=\"Validate a VCF generated by cohort extraction\")\n",
    "#parser.add_argument('--expected', required=True, help=\"A bgzipped VCF. An index file of the same basename must exist in the same directory.\")\n",
    "#parser.add_argument('--actual', required=True, help=\"A bgzipped VCF. An index file of the same basename must exist in the same directory.\")\n",
    "#parser.add_argument('--sample', help=\"The name of one sample for which to validate all genotypes in both VCFs\")\n",
    "#args = parser.parse_args()\n",
    "\n",
    "bcf_bin = \"./bcftools/bcftools\"\n",
    "expected_vcf = \"Merged.vcf.gz\"\n",
    "actual_vcf = \"aou.vcf.gz\"\n",
    "sample_name = None\n",
    "\n",
    "# bcftools stats counts a bunch of stuff in each vcf - records, types of variants,\n",
    "# types of substitution, distribution of variants, etc. We don't care about most\n",
    "# of them but also can't filter down. This handles 1) and 2).\n",
    "expected_stats_proc = subprocess.run([bcf_bin, \"stats\", expected_vcf], capture_output=True, text=True)\n",
    "actual_stats_proc = subprocess.run([bcf_bin, \"stats\", actual_vcf], capture_output=True, text=True)\n",
    "expected_stats = expected_stats_proc.stdout.split(\"\\n\")\n",
    "actual_stats = actual_stats_proc.stdout.split(\"\\n\")\n",
    "\n",
    "# SN stands for Summary Numbers, which covers things like number of samples,\n",
    "# number of records, number of indels, SNPs, multiallelic sites, etc. May as\n",
    "# well check all of them.\n",
    "expected_summary = [stat for stat in expected_stats if stat.startswith('SN')]\n",
    "actual_summary = [stat for stat in actual_stats if stat.startswith('SN')]\n",
    "# We'll grab number of samples for later convenience\n",
    "samples = 0\n",
    "for i in range(0, len(expected_summary)):\n",
    "    expected_stat = expected_summary[i].split(\"\\t\")\n",
    "    actual_stat = actual_summary[i].split(\"\\t\")\n",
    "    if (expected_stat[-1] != actual_stat[-1]):\n",
    "      print(f\"Discrepancy in summary stat '{expected_stat[-2]}': Expected {expected_stat[-1]}, Actual {actual_stat[-2]}\")\n",
    "      sys.exit()\n",
    "    if (actual_stat[2] == \"number of samples:\"):\n",
    "      samples = expected_stat[3]\n",
    "\n",
    "# grab the first sample and pull that sample out of both vcfs.\n",
    "sample_to_check_gts = sample_name if sample_name else None\n",
    "if (sample_to_check_gts is None):\n",
    "    expected_sample_names_proc = subprocess.run([\"bcftools\", \"query\", \"-l\", expected_vcf], capture_output=True, text=True)\n",
    "    expected_sample_names = expected_sample_names_proc.stdout.replace(\"\\'\", \"\").split(\"\\n\")\n",
    "    sample_to_check_gts = expected_sample_names[0]\n",
    "\n",
    "# bcftools query extracts particular fields from each field, such as allele frequency or genotype.\n",
    "expected_query_proc = subprocess.run([bcf_bin, \"query\", \"-f\", \"'%CHROM %POS %AF [%GT]\\n'\", \"-s\", sample_to_check_gts, expected_vcf], capture_output=True, text=True)\n",
    "actual_query_proc = subprocess.run([bcf_bin, \"query\", \"-f\", \"'%CHROM %POS %AF [%GT]\\n'\", \"-s\", sample_to_check_gts, actual_vcf], capture_output=True, text=True)\n",
    "# apPARently there is a newline on the end thus the splice\n",
    "expected_query_stringy = expected_query_proc.stdout.replace(\"\\'\", \"\").split(\"\\n\")[:-1]\n",
    "actual_query_stringy = actual_query_proc.stdout.replace(\"\\'\", \"\").split(\"\\n\")[:-1]\n",
    "expected_af = [af.split() for af in expected_query_stringy]\n",
    "actual_af = [af.split() for af in actual_query_stringy]\n",
    "\n",
    "# munge - map by chromosome then by position.\n",
    "expected_by_chrom_pos = {}\n",
    "for chrom, pos, af, gt in expected_af:\n",
    "    # af is pretty straightforward.\n",
    "    if chrom in expected_by_chrom_pos:\n",
    "      if pos in expected_by_chrom_pos:\n",
    "        expected_by_chrom_pos[chrom][pos]['af'] = af\n",
    "        expected_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "      else:\n",
    "        expected_by_chrom_pos[chrom][pos] = {}\n",
    "        expected_by_chrom_pos[chrom][pos]['af'] = af\n",
    "        expected_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "    else:\n",
    "      expected_by_chrom_pos[chrom] = {}\n",
    "      expected_by_chrom_pos[chrom][pos] = {}\n",
    "      expected_by_chrom_pos[chrom][pos]['af'] = af\n",
    "      expected_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "\n",
    "# and the same for actual.\n",
    "actual_by_chrom_pos = {}\n",
    "for chrom, pos, af, gt in actual_af:\n",
    "    # actual af\n",
    "    if chrom in actual_by_chrom_pos:\n",
    "      if pos in actual_by_chrom_pos:\n",
    "        actual_by_chrom_pos[chrom][pos]['af'] = af\n",
    "        actual_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "      else:\n",
    "        actual_by_chrom_pos[chrom][pos] = {}\n",
    "        actual_by_chrom_pos[chrom][pos]['af'] = af\n",
    "        actual_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "    else:\n",
    "      actual_by_chrom_pos[chrom] = {}\n",
    "      actual_by_chrom_pos[chrom][pos] = {}\n",
    "      actual_by_chrom_pos[chrom][pos]['af'] = af\n",
    "      actual_by_chrom_pos[chrom][pos]['gt'] = gt\n",
    "\n",
    "# chromosomes SHOULD be the same for both of these. we're probably doing them all on humans.\n",
    "for chrom in expected_by_chrom_pos.keys():\n",
    "    expected_positions = sorted(expected_by_chrom_pos[chrom].keys())\n",
    "    actual_positions = sorted(actual_by_chrom_pos[chrom].keys())\n",
    "    if expected_positions != actual_positions:\n",
    "      print(f\"Discrepancy on chromosome {chrom}: sequenced positions are not the same\")\n",
    "      sys.exit()\n",
    "    else:\n",
    "      for pos in expected_positions:\n",
    "        # validate AF. This handles 3)\n",
    "        if expected_by_chrom_pos[chrom][pos]['af'] != actual_by_chrom_pos[chrom][pos]['af']:\n",
    "          print(f\"Discrepancy in allele frequency on chromosome {chrom} at position {pos}: Expected {expected_by_chrom_pos[chrom][pos]['af']}, Actual {actual_af_by_chrom_pos[chrom][pos]['af']}\")\n",
    "          sys.exit()\n",
    "        # validate gt. This handles 4)\n",
    "        if expected_by_chrom_pos[chrom][pos]['gt'] != actual_by_chrom_pos[chrom][pos]['gt']:\n",
    "          print(f\"Discrepancy in genotype on chromosome {chrom} at position {pos}: Expected {expected_by_chrom_pos[chrom][pos]['gt']}, Actual {actual_by_chrom_pos[chrom][pos]['gt']}\")\n",
    "          sys.exit()\n",
    "\n",
    "print(\"All good, this looks like the same set of people.\")\n",
    "end = time.time()\n",
    "print(f\"Elapsed time: {end-start} seconds for {samples} samples\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# This file was produced by bcftools stats (1.10.2-107-gd17d9cd+htslib-1.10.2-117-ga79009b) and can be plotted using plot-vcfstats.\n",
      "# The command line was:\tbcftools stats  aou.vcf.gz\n",
      "#\n",
      "# Definition of sets:\n",
      "# ID\t[2]id\t[3]tab-separated file names\n",
      "ID\t0\taou.vcf.gz\n",
      "# SN, Summary numbers:\n",
      "#   number of records   .. number of data rows in the VCF\n",
      "#   number of no-ALTs   .. reference-only sites, ALT is either \".\" or identical to REF\n",
      "#   number of SNPs      .. number of rows with a SNP\n",
      "#   number of MNPs      .. number of rows with a MNP, such as CC>TT\n",
      "#   number of indels    .. number of rows with an indel\n",
      "#   number of others    .. number of rows with other type, for example a symbolic allele or\n",
      "#                          a complex substitution, such as ACT>TCGA\n",
      "#   number of multiallelic sites     .. number of rows with multiple alternate alleles\n",
      "#   number of multiallelic SNP sites .. number of rows with multiple alternate alleles, all SNPs\n",
      "# \n",
      "#   Note that rows containing multiple types will be counted multiple times, in each\n",
      "#   counter. For example, a row with a SNP and an indel increments both the SNP and\n",
      "#   the indel counter.\n",
      "# \n",
      "# SN\t[2]id\t[3]key\t[4]value\n",
      "SN\t0\tnumber of samples:\t500\n",
      "SN\t0\tnumber of records:\t1900033\n",
      "SN\t0\tnumber of no-ALTs:\t0\n",
      "SN\t0\tnumber of SNPs:\t1856770\n",
      "SN\t0\tnumber of MNPs:\t0\n",
      "SN\t0\tnumber of indels:\t43263\n",
      "SN\t0\tnumber of others:\t0\n",
      "SN\t0\tnumber of multiallelic sites:\t1456\n",
      "SN\t0\tnumber of multiallelic SNP sites:\t1456\n",
      "# TSTV, transitions/transversions:\n",
      "# TSTV\t[2]id\t[3]ts\t[4]tv\t[5]ts/tv\t[6]ts (1st ALT)\t[7]tv (1st ALT)\t[8]ts/tv (1st ALT)\n",
      "TSTV\t0\t1324786\t533440\t2.48\t1324682\t532088\t2.49\n",
      "# SiS, Singleton stats:\n",
      "# SiS\t[2]id\t[3]allele count\t[4]number of SNPs\t[5]number of transitions\t[6]number of transversions\t[7]number of indels\t[8]repeat-consistent\t[9]repeat-inconsistent\t[10]not applicable\n",
      "SiS\t0\t1\t1858226\t1324786\t533440\t43263\t0\t0\t43263\n",
      "# AF, Stats by non-reference allele frequency:\n",
      "# AF\t[2]id\t[3]allele frequency\t[4]number of SNPs\t[5]number of transitions\t[6]number of transversions\t[7]number of indels\t[8]repeat-consistent\t[9]repeat-inconsistent\t[10]not applicable\n",
      "AF\t0\t0.000000\t1858226\t1324786\t533440\t43263\t0\t0\t43263\n",
      "# QUAL, Stats by quality:\n",
      "# QUAL\t[2]id\t[3]Quality\t[4]number of SNPs\t[5]number of transitions (1st ALT)\t[6]number of transversions (1st ALT)\t[7]number of indels\n",
      "QUAL\t0\t998\t1856770\t1324682\t532088\t43263\n",
      "# IDD, InDel distribution:\n",
      "# IDD\t[2]id\t[3]length (deletions negative)\t[4]number of sites\t[5]number of genotypes\t[6]mean VAF\n",
      "IDD\t0\t-60\t236\t0\t.\n",
      "IDD\t0\t-59\t3\t0\t.\n",
      "IDD\t0\t-58\t11\t0\t.\n",
      "IDD\t0\t-57\t12\t0\t.\n",
      "IDD\t0\t-56\t15\t0\t.\n",
      "IDD\t0\t-55\t7\t0\t.\n",
      "IDD\t0\t-54\t17\t0\t.\n",
      "IDD\t0\t-53\t8\t0\t.\n",
      "IDD\t0\t-52\t14\t0\t.\n",
      "IDD\t0\t-51\t8\t0\t.\n",
      "IDD\t0\t-50\t17\t0\t.\n",
      "IDD\t0\t-49\t3\t0\t.\n",
      "IDD\t0\t-47\t3\t0\t.\n",
      "IDD\t0\t-46\t3\t0\t.\n",
      "IDD\t0\t-45\t6\t0\t.\n",
      "IDD\t0\t-44\t4\t0\t.\n",
      "IDD\t0\t-43\t3\t0\t.\n",
      "IDD\t0\t-42\t2\t0\t.\n",
      "IDD\t0\t-41\t5\t0\t.\n",
      "IDD\t0\t-40\t2\t0\t.\n",
      "IDD\t0\t-38\t3\t0\t.\n",
      "IDD\t0\t-37\t2\t0\t.\n",
      "IDD\t0\t-36\t3\t0\t.\n",
      "IDD\t0\t-35\t4\t0\t.\n",
      "IDD\t0\t-34\t1\t0\t.\n",
      "IDD\t0\t-33\t3\t0\t.\n",
      "IDD\t0\t-32\t5\t0\t.\n",
      "IDD\t0\t-30\t9\t0\t.\n",
      "IDD\t0\t-29\t13\t0\t.\n",
      "IDD\t0\t-28\t7\t0\t.\n",
      "IDD\t0\t-27\t27\t0\t.\n",
      "IDD\t0\t-26\t18\t0\t.\n",
      "IDD\t0\t-25\t27\t0\t.\n",
      "IDD\t0\t-24\t33\t0\t.\n",
      "IDD\t0\t-23\t39\t0\t.\n",
      "IDD\t0\t-22\t30\t0\t.\n",
      "IDD\t0\t-21\t43\t0\t.\n",
      "IDD\t0\t-20\t35\t0\t.\n",
      "IDD\t0\t-19\t57\t0\t.\n",
      "IDD\t0\t-18\t62\t0\t.\n",
      "IDD\t0\t-17\t53\t0\t.\n",
      "IDD\t0\t-16\t95\t0\t.\n",
      "IDD\t0\t-15\t106\t0\t.\n",
      "IDD\t0\t-14\t131\t0\t.\n",
      "IDD\t0\t-13\t160\t0\t.\n",
      "IDD\t0\t-12\t199\t0\t.\n",
      "IDD\t0\t-11\t217\t0\t.\n",
      "IDD\t0\t-10\t249\t0\t.\n",
      "IDD\t0\t-9\t268\t0\t.\n",
      "IDD\t0\t-8\t316\t0\t.\n",
      "IDD\t0\t-7\t399\t0\t.\n",
      "IDD\t0\t-6\t389\t0\t.\n",
      "IDD\t0\t-5\t1037\t0\t.\n",
      "IDD\t0\t-4\t3070\t0\t.\n",
      "IDD\t0\t-3\t3654\t0\t.\n",
      "IDD\t0\t-2\t5678\t0\t.\n",
      "IDD\t0\t-1\t14433\t0\t.\n",
      "IDD\t0\t1\t7883\t0\t.\n",
      "IDD\t0\t2\t1232\t0\t.\n",
      "IDD\t0\t3\t436\t0\t.\n",
      "IDD\t0\t4\t829\t0\t.\n",
      "IDD\t0\t5\t210\t0\t.\n",
      "IDD\t0\t6\t85\t0\t.\n",
      "IDD\t0\t7\t74\t0\t.\n",
      "IDD\t0\t8\t94\t0\t.\n",
      "IDD\t0\t9\t51\t0\t.\n",
      "IDD\t0\t10\t39\t0\t.\n",
      "IDD\t0\t11\t27\t0\t.\n",
      "IDD\t0\t12\t18\t0\t.\n",
      "IDD\t0\t13\t27\t0\t.\n",
      "IDD\t0\t14\t26\t0\t.\n",
      "IDD\t0\t15\t9\t0\t.\n",
      "IDD\t0\t16\t12\t0\t.\n",
      "IDD\t0\t17\t12\t0\t.\n",
      "IDD\t0\t18\t8\t0\t.\n",
      "IDD\t0\t19\t24\t0\t.\n",
      "IDD\t0\t20\t8\t0\t.\n",
      "IDD\t0\t21\t6\t0\t.\n",
      "IDD\t0\t22\t4\t0\t.\n",
      "IDD\t0\t23\t3\t0\t.\n",
      "IDD\t0\t24\t5\t0\t.\n",
      "IDD\t0\t26\t9\t0\t.\n",
      "IDD\t0\t28\t5\t0\t.\n",
      "IDD\t0\t29\t1\t0\t.\n",
      "IDD\t0\t31\t3\t0\t.\n",
      "IDD\t0\t34\t1\t0\t.\n",
      "IDD\t0\t39\t2\t0\t.\n",
      "IDD\t0\t40\t3\t0\t.\n",
      "IDD\t0\t42\t2\t0\t.\n",
      "IDD\t0\t43\t6\t0\t.\n",
      "IDD\t0\t48\t1\t0\t.\n",
      "IDD\t0\t50\t9\t0\t.\n",
      "IDD\t0\t51\t6\t0\t.\n",
      "IDD\t0\t52\t8\t0\t.\n",
      "IDD\t0\t53\t6\t0\t.\n",
      "IDD\t0\t54\t10\t0\t.\n",
      "IDD\t0\t55\t6\t0\t.\n",
      "IDD\t0\t56\t7\t0\t.\n",
      "IDD\t0\t57\t3\t0\t.\n",
      "IDD\t0\t58\t3\t0\t.\n",
      "IDD\t0\t59\t6\t0\t.\n",
      "IDD\t0\t60\t790\t0\t.\n",
      "# ST, Substitution types:\n",
      "# ST\t[2]id\t[3]type\t[4]count\n",
      "ST\t0\tA>C\t70116\n",
      "ST\t0\tA>G\t264410\n",
      "ST\t0\tA>T\t46084\n",
      "ST\t0\tC>A\t97949\n",
      "ST\t0\tC>G\t57462\n",
      "ST\t0\tC>T\t396729\n",
      "ST\t0\tG>A\t398254\n",
      "ST\t0\tG>C\t56863\n",
      "ST\t0\tG>T\t89740\n",
      "ST\t0\tT>A\t45619\n",
      "ST\t0\tT>C\t265393\n",
      "ST\t0\tT>G\t69607\n",
      "# DP, Depth distribution\n",
      "# DP\t[2]id\t[3]bin\t[4]number of genotypes\t[5]fraction of genotypes (%)\t[6]number of sites\t[7]fraction of sites (%)\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "./bcftools/bcftools stats aou.vcf.gz"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
